presto-bigquery/src/main/java/com/facebook/presto/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -0,0 +1,236 @@
+    {
+        if (schemaName.isPresent() && schemaName.get().equalsIgnoreCase(INFORMATION_SCHEMA)) {
+            return ImmutableList.of();
+        }
+        Set<String> schemaNames = schemaName.map(ImmutableSet::of)
+                .orElseGet(() -> ImmutableSet.copyOf(listSchemaNames(session)));
+
+        ImmutableList.Builder<SchemaTableName> tableNames = ImmutableList.builder();
+        for (String datasetId : schemaNames) {
+            for (Table table : bigQueryClient.listTables(DatasetId.of(projectId, datasetId), types)) {
+                tableNames.add(new SchemaTableName(datasetId, table.getTableId().getTable()));
+            }
+        }
+        return tableNames.build();
+    }
+
+    @Override
+    public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        log.debug("getTableHandle(session=%s, tableName=%s)", session, tableName);
+        TableInfo tableInfo = getBigQueryTable(tableName);
+        if (tableInfo == null) {
+            log.debug("Table [%s.%s] was not found", tableName.getSchemaName(), tableName.getTableName());
+            return null;
+        }
+        return BigQueryTableHandle.from(tableInfo);
+    }
+
+    @Override
+    public List<ConnectorTableLayoutResult> getTableLayouts(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            Constraint<ColumnHandle> constraint,
+            Optional<Set<ColumnHandle>> desiredColumns)
+    {
+        log.debug("getTableMetadata(session=%s, table=%s, constraint=%s, desiredColumns=%s)", session, table, constraint, desiredColumns);
+        BigQueryTableHandle bigQueryTableHandle = (BigQueryTableHandle) table;
+        if (desiredColumns.isPresent()) {
+            bigQueryTableHandle = bigQueryTableHandle.withProjectedColumns(ImmutableList.copyOf(desiredColumns.get()));
+        }
+        BigQueryTableLayoutHandle bigQueryTableLayoutHandle = new BigQueryTableLayoutHandle(bigQueryTableHandle);
+        return ImmutableList.of(new ConnectorTableLayoutResult(new ConnectorTableLayout(bigQueryTableLayoutHandle), constraint.getSummary()));
+    }
+
+    @Override
+    public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle layoutHandle)
+    {
+        log.debug("getTableMetadata(session=%s, layoutHandle=%s)", session, layoutHandle);
+        BigQueryTableLayoutHandle bigQueryTableLayoutHandle = (BigQueryTableLayoutHandle) layoutHandle;
+        return new ConnectorTableLayout(
+                bigQueryTableLayoutHandle,
+                Optional.empty(), // columns of the table, not projected
+                bigQueryTableLayoutHandle.getTupleDomain(), // predicate
+                Optional.empty(), // tablePartitioning
+                Optional.empty(), // streamPartitioningColumns
+                Optional.empty(), // discretePredicates
+                ImmutableList.of()); // localProperties
+    }
+
+    // May return null
```

Comment:
could we try return Optional<TableInfo> if possible? we try to get rid of return null everywhere in the code

Reply:
ok wrap by Optional now